kedro/runner/runner.py

```diff
@@ -90,11 +95,11 @@ def run(
         # Check if there's any output datasets that aren't in the catalog and don't match a pattern
         # in the catalog.
         free_outputs = pipeline.outputs() - set(registered_ds)
-        unregistered_ds = pipeline.datasets() - set(registered_ds)
 
-        # Create a default dataset for unregistered datasets
-        for ds_name in unregistered_ds:
-            catalog.add(ds_name, self.create_default_dataset(ds_name))
+        # Register the default dataset pattern with the catalog
+        catalog = catalog.shallow_copy(
+            extra_dataset_patterns=self._extra_dataset_patterns
+        )
```

Comment:
If this is moved up we can't accurately determine the `unsatisfied` inputs that aren't defined in the catalog and don't match a pattern or the `free_outputs` because both would be captured under the default dataset which isn't supposed to happen at that stage yet.